File: Java/src/main/java/QuestionBank/Solution1001_1100/Solution1005.java
package QuestionBank.Solution1001_1100;

import java.util.ArrayList;
import java.util.Collections;

/**
 * @author windows
 */
public class Solution1005 {
    public int largestSumAfterKNegations(int[] nums, int k) {
        ArrayList<Integer> list = new ArrayList<>();
        int min = Integer.MAX_VALUE;
        int sum = 0;
        for (int i = 0; i < nums.length; i++) {
            if(nums[i] < 0){
                list.add(nums[i]);
            }
            min = Math.min(min, Math.abs(nums[i]));
            sum += nums[i];
        }
        Collections.sort(list);
        if (k <= list.size()){
            for (int i = 0; i < k; i++) {
                sum += 2 * Math.abs(list.get(i));
            }
        }else{
            for (int i = 0; i < list.size(); i++) {
                sum += 2 * Math.abs(list.get(i));
            }
            boolean flag = (k - list.size()) % 2 == 0;
            if (!flag){
                sum -= min * 2;
            }
        }
        return sum;
    }
    
    
    public static void main(String[] args) {
        Solution1005 solution1005 = new Solution1005();
        System.out.println(solution1005.largestSumAfterKNegations(new int[]{3, -1, 0 ,2}, 3));
    }
}

File: C++/src/Solution1001_1100/Solution1005.cpp
//
// Created by windows on 2023/3/2.
//

#include "Solution1005.h"
#include <functional>
#include <algorithm>
#include <iostream>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>
using namespace std;

int largestSumAfterKNegations(vector<int>& nums, int k) {
  sort(nums.begin(), nums.end());
  int n = nums.size();

  for (int i = 0; i < nums.size(); ++i) {
    if (nums[i] > 0 || k <= 0){
      break;
    }
    nums[i] = -nums[i];
    k--;
  }
  sort(nums.begin(), nums.end());

  if (k % 2== 1){
    nums[0] = -nums[0];
  }

  int sum = 0;
  for (int i = 0; i < nums.size(); ++i) {
    sum += nums[i];
  }
  return sum;
}
